### ISSUE TITLE
Cluster: Download with dwc fields failing 

### ISSUE THREAD

**Issue by javier-molina** [AtlasOfLivingAustralia/biocache-service#208]
A download using Full Darwin Core is failing as the fields call return duplicate field definition entries that are missing the downloadName attribute.

A call to https://biocache-clustered.ala.org.au/ws/index/fields returnts roughly 4100+ entries.

A couple of offending instances are below:

```
{
"name": "county",
"dataType": "string",
"indexed": false,
"stored": false,
"multivalue": false,
"info": "http://rs.tdwg.org/dwc/terms/county",
"dwcTerm": "county",
"downloadName": "county",
"classs": "Location"
},
{
"name": "county",
"dataType": "string",
"indexed": true,
"stored": true,
"multivalue": false,
"info": "http://rs.tdwg.org/dwc/terms/county",
"dwcTerm": "county",
"classs": "Location"
},

```
and 
```

{
"name": "day",
"dataType": "string",
"indexed": false,
"stored": false,
"multivalue": false,
"info": "http://rs.tdwg.org/dwc/terms/day",
"dwcTerm": "day_raw",
"downloadName": "day",
"classs": "Event"
},
{
"name": "day",
"dataType": "string",
"indexed": true,
"stored": true,
"multivalue": false,
"info": "http://rs.tdwg.org/dwc/terms/day",
"dwcTerm": "day_raw",
"classs": "Event"
},
```

in both instances the dwcTerm is present,  [hinting downloads plugin](https://github.com/AtlasOfLivingAustralia/downloads-plugin/blob/3eca07f5cbe75bb44f203eab0ba563f6d3ae20e7/grails-app/services/au/org/ala/downloads/BiocacheService.groovy#L94)  to include the field in the download however one of those entries lacks the downloadName attribute which eventually results in the error below in biocache-service

```
2018-04-16 10:25:43,236 WARN : (DownloadService.java:1258) - Offline download failed. Cassandra driver error. Retrying in 5 mins. Task file: /data/biocache-download/b8440880-a800-3e0a-b710-a8aecca1da90/1523838341767/records-2018-04-16.zip
com.datastax.driver.core.exceptions.InvalidQueryException: Undefined column name "null"
        at com.datastax.driver.core.exceptions.InvalidQueryException.copy(InvalidQueryException.java:50)
        at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:37)
        at com.datastax.driver.core.AbstractSession.prepare(AbstractSession.java:104)
        at au.org.ala.biocache.persistence.Cassandra3PersistenceManager.selectRows(Cassandra3PersistenceManager.scala:1130)
        at au.org.ala.biocache.dao.OccurrenceDAOImpl.writeToRecordWriter(OccurrenceDAOImpl.scala:221)
        at au.org.ala.biocache.Store$.writeToWriter(Store.scala:537)
        at au.org.ala.biocache.Store.writeToWriter(Store.scala)
        at au.org.ala.biocache.dao.SearchDAOImpl.downloadRecords(SearchDAOImpl.java:1990)
        at au.org.ala.biocache.dao.SearchDAOImpl.writeResultsToStream(SearchDAOImpl.java:1775)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:198)
        at com.sun.proxy.$Proxy30.writeResultsToStream(Unknown Source)
        at au.org.ala.biocache.service.DownloadService.writeQueryToStream(DownloadService.java:538)
        at au.org.ala.biocache.service.DownloadService$DownloadCreatorImpl$1.call(DownloadService.java:1188)
        at au.org.ala.biocache.service.DownloadService$DownloadCreatorImpl$1.call(DownloadService.java:1159)
        at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
        at java.util.concurrent.FutureTask.run(FutureTask.java)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: Undefined column name "null"
        at com.datastax.driver.core.Responses$Error.asException(Responses.java:147)
        at com.datastax.driver.core.SessionManager$4.apply(SessionManager.java:220)
        at com.datastax.driver.core.SessionManager$4.apply(SessionManager.java:196)
        at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:211)
        at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.doTransform(AbstractTransformFuture.java:200)
        at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:111)
        at com.google.common.util.concurrent.MoreExecutors$5$1.run(MoreExecutors.java:992)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        ... 1 more
```

It is also worth mentioning that a  call to https://biocache-clustered.ala.org.au/ws/index/fields returnts roughly 4100+ entries 
while https://biocache.ala.org.au/ws/index/fields returns 900+ only

**Comment by ansell** [comment]
Code relevant to the issue may be:

https://github.com/AtlasOfLivingAustralia/biocache-service/blob/66f7b1fedca5ac5023ce2d069299738d5f04a7f7/src/main/java/au/org/ala/biocache/dao/SearchDAOImpl.java#L3201...L3202

Also has some other strange results such as:

```json
  {
    "name": "",
    "dataType": "null",
    "indexed": false,
    "stored": false,
    "multivalue": false,
    "downloadName": "zone"
  },
  {
    "name": "_",
    "dataType": "null",
    "indexed": false,
    "stored": false,
    "multivalue": false
  }
```

**Comment by javier-molina** [comment]
[indexFields-bstore-1.9.3.txt](https://github.com/AtlasOfLivingAustralia/biocache-service/files/1918638/indexFields-bstore-1.9.3.txt)
[indexFields-bstore-2.3.txt](https://github.com/AtlasOfLivingAustralia/biocache-service/files/1918639/indexFields-bstore-2.3.txt)




[solr-b4-luke-fields.xml.txt](https://github.com/AtlasOfLivingAustralia/biocache-service/files/1918592/solr-b4-luke-fields.xml.txt)

**Comment by javier-molina** [comment]
@adam-collins the number of fields went down from 4000+ to 1000+ yet the same error remains re: duplicate fields one without downloadName, the same two example records are still showing.

Have a look to commit https://github.com/AtlasOfLivingAustralia/biocache-service/commit/b9220e8ee91561d2912bf46ee2872da5f42352d3 I added the need it filter, let me know if you think I'm missing anything.

**Comment by javier-molina** [comment]
And still failing with missing column 'references' in Cassandra. 
Below is the current definition of table occ. 

```
cqlsh:occ> DESCRIBE TABLE occ.occ;
CREATE TABLE occ.occ (
    rowkey text PRIMARY KEY,
    "IdentificationQualifierProcessor_qa" text,
    "_class" text,
    "_qa" text,
    "abcdIdentificationQualifier" text,
    "abcdIdentificationQualifierInsertionPoint" text,
    "abcdIdentificationQualifier_p" text,
    "abcdTypeStatus" text,
    "acceptedNameUsage" text,
    "acceptedNameUsageID" text,
    "accessRights" text,
    "alaUserId" text,
    "associatedMedia" text,
    "associatedOccurrences" text,
    "associatedOccurrences_p" text,
    "associatedReferences" text,
    "associatedReferences_p" text,
    "associatedSequences" text,
    "associatedTaxa" text,
    attr_qa text,
    "austConservation_p" text,
    "australianHerbariumRegion" text,
    "basisOfRecord" text,
    "basisOfRecord_p" text,
    bbox text,
    bbox_p text,
    behavior text,
    "bibliographicCitation" text,
    biome text,
    biome_p text,
    bor_qa text,
    "catalogNumber" text,
    citation text,
    cl_p text,
    class text,
    "classID" text,
    "classID_p" text,
    class_p text,
    class_qa text,
    classs text,
    classs_p text,
    "collectionCode" text,
    "collectionCode_p" text,
    "collectionID" text,
    "collectionName" text,
    "collectionName_p" text,
    "collectionUid" text,
    "collectionUid_p" text,
    "collectorFieldNumber" text,
    continent text,
    continent_p text,
    "coordinatePrecision" text,
    "coordinatePrecision_p" text,
    "coordinateUncertaintyInMeters" text,
    "coordinateUncertaintyInMeters_p" text,
    country text,
    "countryCode" text,
    "countryCode_p" text,
    "countryConservation" text,
    "countryConservation_p" text,
    country_p text,
    county text,
    "cultivarName" text,
    cultivated text,
    "dataGeneralizations" text,
    "dataGeneralizations_p" text,
    "dataHubName" text,
    "dataHubUid" text,
    "dataHubUid_p" text,
    "dataProviderName" text,
    "dataProviderName_p" text,
    "dataProviderUid" text,
    "dataProviderUid_p" text,
    "dataResourceName" text,
    "dataResourceName_p" text,
    "dataResourceUid" text,
    "datasetID" text,
    "datasetName" text,
    "dateDeleted" text,
    "dateIdentified" text,
    "dateIdentified_p" text,
    "datePrecision" text,
    "datePrecision_p" text,
    day text,
    day_p text,
    "decimalLatitude" text,
    "decimalLatitude_p" text,
    "decimalLatitudelatitude" text,
    "decimalLongitude" text,
    "decimalLongitude_p" text,
    "defaultValuesUsed" text,
    default_qa text,
    deleted text,
    disposition text,
    "distanceOutsideExpertRange" text,
    "distanceOutsideExpertRange_p" text,
    duplicates text,
    "duplicatesOriginalInstitutionID" text,
    "duplicatesOriginalUnitID" text,
    duplicates_qa text,
    "duplicationStatus" text,
    "duplicationStatus_p" text,
    "duplicationType" text,
    "duplicationType_p" text,
    "dynamicProperties" text,
    easting text,
    easting_p text,
    el_p text,
    "endDayOfYear" text,
    "endYear" text,
    "establishmentMeans" text,
    "establishmentMeans_p" text,
    "eventAttributes" text,
    "eventDate" text,
    "eventDateEnd" text,
    "eventDateEnd_p" text,
    "eventDate_p" text,
    "eventID" text,
    "eventRemarks" text,
    "eventTime" text,
    event_qa text,
    family text,
    "familyID" text,
    "familyID_p" text,
    family_p text,
    "fieldNotes" text,
    "fieldNumber" text,
    "firstLoaded" text,
    "footprintSRS" text,
    "footprintSpatialFit" text,
    "footprintWKT" text,
    "generalisationToApplyInMetres" text,
    "generalisedLocality" text,
    genus text,
    "genusID" text,
    "genusID_p" text,
    genus_p text,
    "geodeticDatum" text,
    "geodeticDatum_p" text,
    "georeferenceProtocol" text,
    "georeferenceProtocol_p" text,
    "georeferenceRemarks" text,
    "georeferenceSources" text,
    "georeferenceSources_p" text,
    "georeferenceVerificationStatus" text,
    "georeferenceVerificationStatus_p" text,
    "georeferencedBy" text,
    "georeferencedBy_p" text,
    "georeferencedDate" text,
    "georeferencedDate_p" text,
    "geospatialIssue" text,
    "geospatiallyKosher" text,
    "globalConservation" text,
    "gridReference" text,
    habitat text,
    habitat_p text,
    "higherClassification" text,
    "higherGeography" text,
    "higherGeographyID" text,
    ibra text,
    "ibraSubregion" text,
    ibra_p text,
    "identificationID" text,
    "identificationQualifier" text,
    "identificationQualifier_p" text,
    "identificationReferences" text,
    "identificationReferences_p" text,
    "identificationRemarks" text,
    "identificationVerificationStatus" text,
    "identificationVerificationStatus_p" text,
    "identifiedBy" text,
    "identifiedBy_p" text,
    identifier text,
    "identifierBy" text,
    "identifierRole" text,
    image_qa text,
    images text,
    images_p text,
    imcra_p text,
    "individualCount" text,
    "individualID" text,
    "informationWithheld" text,
    "informationWithheld_p" text,
    "infraspecificEpithet" text,
    "infraspecificMarker" text,
    "institutionCode" text,
    "institutionCode_p" text,
    "institutionID" text,
    "institutionID_p" text,
    "institutionName" text,
    "institutionName_p" text,
    "institutionUid" text,
    "institutionUid_p" text,
    interactions text,
    interactions_p text,
    island text,
    "islandGroup" text,
    kingdom text,
    "kingdomID" text,
    "kingdomID_p" text,
    kingdom_p text,
    language text,
    "lastModifiedTime" text,
    "lastModifiedTime_p" text,
    "lastUserAssertionDate" text,
    left text,
    left_p text,
    lga text,
    lga_p text,
    license text,
    license_p text,
    "lifeStage" text,
    "lifeStage_p" text,
    "loanDate" text,
    "loanDestination" text,
    "loanForBotanist" text,
    "loanIdentifier" text,
    "loanReturnDate" text,
    "loanSequenceNumber" text,
    loc_qa text,
    locality text,
    locality_p text,
    "locationAccordingTo" text,
    "locationAttributes" text,
    "locationDetermined" text,
    "locationID" text,
    "locationRemarks" text,
    "maximumDepthInMeters" text,
    "maximumDepthInMeters_p" text,
    "maximumDistanceAboveSurfaceInMeters" text,
    "maximumElevationInMeters" text,
    "maximumElevationInMeters_p" text,
    "measurementAccuracy" text,
    "measurementDeterminedBy" text,
    "measurementDeterminedDate" text,
    "measurementID" text,
    "measurementMethod" text,
    "measurementRemarks" text,
    "measurementType" text,
    "measurementUnit" text,
    "measurementValue" text,
    "minimumDepthInMeters" text,
    "minimumDepthInMeters_p" text,
    "minimumDistanceAboveSurfaceInMeters" text,
    "minimumElevationInMeters" text,
    "minimumElevationInMeters_p" text,
    "miscProperties" text,
    modified text,
    modified_p text,
    month text,
    month_p text,
    municipality text,
    mytest text,
    "nameAccordingTo" text,
    "nameAccordingToID" text,
    "nameAccordingTo_p" text,
    "nameMatchMetric" text,
    "nameMatchMetric_p" text,
    "nameParseType" text,
    "nameParseType_p" text,
    "namePublishedIn" text,
    "namePublishedInID" text,
    "namePublishedInYear" text,
    "naturalOccurrence" text,
    "nearNamedPlaceRelationTo" text,
    "nomenclaturalCode" text,
    "nomenclaturalCode_p" text,
    "nomenclaturalStatus" text,
    northing text,
    northing_p text,
    "occurrenceAttributes" text,
    "occurrenceDetails" text,
    "occurrenceID" text,
    "occurrenceRemarks" text,
    "occurrenceStatus" text,
    "occurrenceStatus_p" text,
    offline_qa text,
    "order" text,
    "orderID" text,
    "orderID_p" text,
    order_p text,
    "originalDecimalLatitude" text,
    "originalDecimalLongitude" text,
    "originalNameUsage" text,
    "originalNameUsageID" text,
    "originalSensitiveValues" text,
    "otherCatalogNumbers" text,
    "outlierForLayers" text,
    "outlierForLayers_p" text,
    "ownerInstitutionCode" text,
    "parentNameUsage" text,
    "parentNameUsageID" text,
    phenology text,
    "photoPageUrl" text,
    photographer text,
    phylum text,
    "phylumID" text,
    "phylumID_p" text,
    phylum_p text,
    "pointRadiusSpatialFit" text,
    "portalId" text,
    "preferredFlag" text,
    preparations text,
    "previousIdentifications" text,
    provenance text,
    provenance_p text,
    "qualityAssertion" text,
    "queryAssertions_p" text,
    "recordNumber" text,
    "recordedBy" text,
    "recordedBy_p" text,
    "relatedResourceID" text,
    "relationshipAccordingTo" text,
    "relationshipEstablishedDate" text,
    "relationshipOfResource" text,
    "relationshipRemarks" text,
    reprocessing_qa text,
    "reproductiveCondition" text,
    "resourceID" text,
    "resourceRelationshipID" text,
    right text,
    right_p text,
    rights text,
    rightsholder text,
    "samplingEffort" text,
    "samplingProtocol" text,
    "samplingProtocol_p" text,
    "scientificName" text,
    "scientificNameAddendum" text,
    "scientificNameAuthorship" text,
    "scientificNameID" text,
    "scientificNameWithoutAuthor" text,
    "scientificName_p" text,
    "secondaryCollectors" text,
    sensitive_qa text,
    sex text,
    sounds text,
    sounds_p text,
    source text,
    species text,
    "speciesGroups" text,
    "speciesGroups_p" text,
    "speciesHabitats" text,
    "speciesHabitats_p" text,
    "speciesID" text,
    "speciesID_p" text,
    species_p text,
    "specificEpithet" text,
    "startDayOfYear" text,
    "startYear" text,
    state text,
    "stateConservation" text,
    "stateConservation_p" text,
    "stateProvince" text,
    "stateProvince_p" text,
    subfamily text,
    subgenus text,
    "subgenusID" text,
    subspecies text,
    "subspeciesID" text,
    "subspeciesID_p" text,
    subspecies_p text,
    superfamily text,
    "taxonConceptID" text,
    "taxonConceptID_p" text,
    "taxonID" text,
    "taxonRank" text,
    "taxonRankID" text,
    "taxonRankID_p" text,
    "taxonRank_p" text,
    "taxonRemarks" text,
    "taxonomicIssue" text,
    "taxonomicIssue_p" text,
    "taxonomicStatus" text,
    "taxonomicallyKosher" text,
    type text,
    "typeStatus" text,
    "typeStatusQualifier" text,
    "typeStatus_p" text,
    type_qa text,
    "typifiedName" text,
    "userAssertionStatus" text,
    "userId" text,
    "userId_p" text,
    "userQualityAssertion" text,
    "userVerified" text,
    "validDistribution" text,
    "verbatimCoordinateSystem" text,
    "verbatimCoordinates" text,
    "verbatimDateIdentified" text,
    "verbatimDepth" text,
    "verbatimDepth_p" text,
    "verbatimElevation" text,
    "verbatimElevation_p" text,
    "verbatimEventDate" text,
    "verbatimLatitude" text,
    "verbatimLocality" text,
    "verbatimLongitude" text,
    "verbatimSRS" text,
    "verbatimTaxonRank" text,
    "verificationDate" text,
    "verificationNotes" text,
    verifier text,
    "vernacularName" text,
    "vernacularName_p" text,
    videos text,
    videos_p text,
    "waterBody" text,
    year text,
    year_p text,
    zone text
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy', 'tombstone_compaction_interval': '18000', 'tombstone_threshold': '0.1'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
```

**Comment by javier-molina** [comment]
The file [Missing fields in Cassandra.txt](https://github.com/AtlasOfLivingAustralia/biocache-service/files/1959249/Missing.fields.in.Cassandra.txt) contains the fields that are not in cassandra cluster set up and that were removed from https://github.com/AtlasOfLivingAustralia/biocache-store/blob/585b46d65e34d162d0313aa4ceaa059f5b9cce39/src/main/resources/indexFields.txt#L1

**Comment by ansell** [comment]
Most of those fields (almost all) are standardised Darwin Core Terms:


https://wayback.archive.org/web/20180404045244/http://rs.tdwg.org/dwc/terms/



________________________________
________________________________

**Comment by djtfmartin** [comment]
this is now fixed with #213
